### PR TITLE
[lab/login] please add federation

### PR DESCRIPTION
**Pull request by Svenlaa**
i hope this code is fine

#### todo
- [ ] add user-styles.css file for other instances

#### federates with
- { name: 'todepond.com', url: 'https://todepond-lablogingetusers.web.val.run' },
- { name: 'svenlaa.com', url: 'https://svenlaa-lablogingetusers.web.val.run' },
- { name: 'evolved.systems', url: 'https://evol-lablogingetusers.web.val.run' },
- { name: 'rossilaz.xyz', url: 'https://mittzy-loginredux_getusers.web.val.run' },

#### how?
1. request all those URLs
2. add the reponse objects to the entries array, including the new instance name
3. act like everything is fine

#### other
- assumes the first instance in the list in SELF instance
- loads user-styles.css file from svenlaa.com for their styles
- usernames get a lot longer, will break `PodeTond 💚` style
- puts instance-specific things into `const`s at the top of the javascript so it can easily be found when making a new instance
- please review the code, suggest changes. i am bad at big picture stuff

### CODE REVIEW

File: lab/login/index.html [hunk 1]
```diff
@@ -61,6 +66,32 @@ <h2 id="feed-heading">
...
+    },
+  ];
+  const PAGE_PATH = "/lab/login/";
+  const ADMIN_NAME = "TodePond";
+  const DELETE_ACCOUNT_URL =
+    "https://todepond-lablogindeleteaccount.web.val.run";
+  const UPDATE_STATUS_URL = "https://todepond-labloginupdatestatus.web.val.run";
+  const BAN_USER_URL = "https://todepond-labloginbanuser.web.val.run";
+  const LOGIN_URL = "https://todepond-lablogin.web.val.run";
+
```
[the remark this answers is not in the document]
Comment (TodePond) [reply lab/login/index.html:94]: and probably have them as a big global JSON object

File: lab/login/index.html [hunk 2]
```diff
@@ -323,7 +350,7 @@ <h2 id="feed-heading">
     } catch (e) {
       users = [];
     }
-    const usernames = users.map((v) => v.username);
+    const usernames = users.map((v) => v.username + "@" + v.instance);
```
Comment (TodePond) [inline lab/login/index.html:353]: okay i think we should probably keep track of aliases here. each user of the home server could also be referred to by *just* their username

File: lab/login/index.html [hunk 3]
```diff
@@ -411,26 +438,27 @@ <h2 id="feed-heading">
...
+      status.className = "post-" + entry.username + "@" + entry.instance;
       status.classList.add("status");
+      status.classList.add("instance-" + entry.instance);
       const strong = status.querySelector("strong");
-      status.id = "post-" + entry.username;
-      strong.className = "name-" + entry.username;
-      strong.textContent = entry.username;
+      status.id = "post-" + entry.username + "@" + entry.instance;
+      strong.className = "name-" + entry.username + "@" + entry.instance;
+      strong.textContent = entry.username + "@" + entry.instance;
```
Comment (TodePond) [inline lab/login/index.html:447]: in a couple of places, i think we could display the instance separately from the name, or not at all. eg: like a subtle little tag

File: lab/login/style.css [hunk 4]
```diff
@@ -318,7 +328,7 @@ input[type="checkbox"] {
   font-family: "JetBrains Mono", sans-serif;
 }
 
-.at-PodeTond.💚:hover {
+.at-PodeTond.💚\@todepond\.com:hover {
```
Comment (TodePond) [inline lab/login/style.css:331]: i think we can fix these ones, even by hardcoding maybe

File: lab/login/index.html [hunk 5]
```diff
@@ -14,6 +15,10 @@
 <meta name="twitter:image" content="https://todepond.com/lab/og.png" />
 
 <link rel="stylesheet" href="/style.css#v2" />
+<link
+  rel="stylesheet"
+  href="https://svenlaa.com/playground/loggo/user-styles.css"
+/>
```
[the remark this answers is not in the document]
Comment (TodePond) [reply lab/login/index.html:21]: or maybe not, i don't know

**Comment by tkukec** [comment]
>     * usernames get a lot longer, will break `PodeTond 💚` style

The code was pretty bad anyways (and browser dependent whoch sucks so much). Feel free to remove it / roll it back to gradient only, and I'll figure out something better

**Comment by TodePond** [comment]
> > ```
> > * usernames get a lot longer, will break `PodeTond 💚` style
> > ```
> 
> The code was pretty bad anyways (and browser dependent whoch sucks so much). Feel free to remove it / roll it back to gradient only, and I'll figure out something better

it's okay should be do-able :)

**Comment by TodePond** [comment]
check out Logarithm, which seems to have done a lot of these changes! 🤯

https://tristie.org/logarithm/

File: lab/login/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (TodePond) [reply lab/login/index.html:71]: nice!

File: lab/login/index.html [hunk 6]
```diff
@@ -110,57 +141,57 @@ <h2 id="feed-heading">
       const namey = document.createElement("span");
       // <span class="name-${username}">${username}</span>!`;
       namey.textContent = username;
-      namey.className = `name-${username}`;
+      namey.className = `name-${username}@${INSTANCES[0].name}`;
```
[the remark this answers is not in the document]
Comment (TodePond) [reply lab/login/index.html]: nice!

File: lab/login/index.html [hunk 7]
```diff
@@ -291,25 +316,27 @@ <h2 id="feed-heading">
...
 
     const entries = [];
+    for (const instance of INSTANCES) {
+      const response = await fetch(instance.url);
+      const json = await response.json();
+      for (const user of json) {
+        const [status, username, timestamp, gif] = user;
+        if (new Date(timestamp) > new Date()) {
+          continue;
+        }
```
[the remark this answers is not in the document]
Comment (TodePond) [reply lab/login/index.html:311]: nice!

**Comment by TodePond** [comment]
awesome stuff, I marked as resolved the ones that I think are done :)

i think it's almost there. probably just needs replies/mentions fixing for when you @ someone on your home server without specifying host - then it should implicitly refer to the person on your server. this is for backwards compatible with existing logs

if i have time I'll do it but anyone else is welcome to jump in 🙏

File: lab/login/index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (TodePond) [reply lab/login/index.html:21]: this is a nit: please ignore

File: lab/login/index.html (hunk 3 above)
Comment (TodePond) [reply lab/login/index.html:428]: this is a nit: please ignore

File: lab/login/style.css (hunk 4 above)
Comment (TodePond) [reply lab/login/style.css:331]: this is a nit: please ignore

File: lab/login/index.html (hunk 2 above)
Comment (TodePond) [reply lab/login/index.html:334]: (last one before merge!!) ✈️

**Comment by TodePond** [comment]
yolo

**Pull request merged**